runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobServerDriver.java

```diff
@@ -0,0 +1,198 @@
+      try {
+        artifactStagingServer.close();
+        LOG.info(
+            "ArtifactStagingServer stopped on {}",
+            artifactStagingServer.getApiServiceDescriptor().getUrl());
+        artifactStagingServer = null;
+      } catch (Exception e) {
+        LOG.error("Error while closing the artifactStagingServer.", e);
+      }
+    }
+  }
+
+  private GrpcFnServer<InMemoryJobService> createJobServer() throws IOException {
+    InMemoryJobService service = createJobService();
+    GrpcFnServer<InMemoryJobService> jobServiceGrpcFnServer;
+    if (configuration.port == 0) {
+      jobServiceGrpcFnServer = GrpcFnServer.allocatePortAndCreateFor(service, jobServerFactory);
+    } else {
+      Endpoints.ApiServiceDescriptor descriptor =
+          Endpoints.ApiServiceDescriptor.newBuilder()
+              .setUrl(configuration.host + ":" + configuration.port)
+              .build();
+      jobServiceGrpcFnServer = GrpcFnServer.create(service, descriptor, jobServerFactory);
+    }
+    LOG.info("JobService started on {}", jobServiceGrpcFnServer.getApiServiceDescriptor().getUrl());
+    return jobServiceGrpcFnServer;
+  }
+
+  private InMemoryJobService createJobService() throws IOException {
+    artifactStagingServer = createArtifactStagingService();
+    JobInvoker invoker = createJobInvoker();
+    return InMemoryJobService.create(
+        artifactStagingServer.getApiServiceDescriptor(),
+        (String session) -> {
+          try {
+            return BeamFileSystemArtifactStagingService.generateStagingSessionToken(
+                session, configuration.artifactStagingPath);
+          } catch (Exception exn) {
+            throw new RuntimeException(exn);
+          }
+        },
+        (String stagingSessionToken) -> {
+          if (configuration.cleanArtifactsPerJob) {
+            artifactStagingServer.getService().removeArtifacts(stagingSessionToken);
+          }
+        },
+        invoker);
+  }
+
+  private GrpcFnServer<BeamFileSystemArtifactStagingService> createArtifactStagingService()
+      throws IOException {
+    BeamFileSystemArtifactStagingService service = new BeamFileSystemArtifactStagingService();
+    final GrpcFnServer<BeamFileSystemArtifactStagingService> artifactStagingService;
+    if (configuration.artifactPort == 0) {
+      artifactStagingService =
+          GrpcFnServer.allocatePortAndCreateFor(service, artifactServerFactory);
+    } else {
+      Endpoints.ApiServiceDescriptor descriptor =
+          Endpoints.ApiServiceDescriptor.newBuilder()
+              .setUrl(configuration.host + ":" + configuration.artifactPort)
+              .build();
+      artifactStagingService = GrpcFnServer.create(service, descriptor, artifactServerFactory);
+    }
+    LOG.info(
+        "ArtifactStagingService started on {}",
+        artifactStagingService.getApiServiceDescriptor().getUrl());
+    return artifactStagingService;
+  }
+
+  protected abstract JobInvoker createJobInvoker();
```

Comment:
Perhaps move this method up to the top so it's easier for users to see what they need to implement.

Reply:
makes sense 👍